bundles/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/internal/handler/EnergenieHandler.java

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.energenie.internal.handler;
+
+import static org.openhab.binding.energenie.internal.EnergenieBindingConstants.*;
+
+import java.io.IOException;
+import java.net.UnknownHostException;
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.cache.ExpiringCache;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.util.ThingHandlerHelper;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.openhab.binding.energenie.internal.config.EnergenieConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link EnergenieHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Hans-Jörg Merk - Initial contribution
+ */
+@NonNullByDefault
+public class EnergenieHandler extends BaseThingHandler {
+
+    private static final String CHANNEL_SOCKET_PREFIX = "socket";
+    private final Logger logger = LoggerFactory.getLogger(EnergenieHandler.class);
+
+    /**
+     * Use cache for refresh command to not update again when call is made within 4 seconds of previous call.
+     */
+    private final ExpiringCache<Boolean> refreshCache = new ExpiringCache<>(Duration.ofSeconds(5), this::refreshState);
+    private final String statusOn;
+
+    private @Nullable EnergenieSocket energenieSocket;
+    private @Nullable ScheduledFuture<?> refreshJob;
+
+    private int refreshInterval;
+    private @Nullable String host;
+
+    public EnergenieHandler(final Thing thing, final String protocol) {
+        super(thing);
+        switch (protocol) {
+            case "EG_PROTO_V20":
+                statusOn = STATE_ON;
+                break;
+            case "EG_PROTO_V21":
+                statusOn = V21_STATE_ON;
+                break;
+            case "EG_PROTO_WLAN":
+                statusOn = WLAN_STATE_ON;
+                break;
+            default:
+                statusOn = STATE_ON;
+                break;
+        }
```

Comment:
This is what @J-N-K was recommending:

Create an enum class:
```
@NonNullByDefault
public enum Protocol{
V20(STATE_ON),
V21(V21_STATE_ON),
WLAN(WLAN_STATE_ON);

private final String statusOn;

private Protocol(String statusOn){
    this.statusOn = statusOn;
}

public String getStatusOn(){
    return statusOn;
}
}
```

Then you can change this code
```suggestion
    public EnergenieHandler(final Thing thing, final Protocol protocol) {
        super(thing);
        this.statusOn = protocol.getStatusOn();
```

And finally you would change your handler factory to this:
```
        if (THING_TYPE_PMSLAN.equals(thingTypeUID)) {
            return new EnergenieHandler(thing, Protocol.V20);
        } else if (THING_TYPE_PM2LAN.equals(thingTypeUID) || THING_TYPE_PMS2LAN.equals(thingTypeUID)) {
            return new EnergenieHandler(thing, Protocol.V21);
        } else if (THING_TYPE_PMSWLAN.equals(thingTypeUID)) {
            return new EnergenieHandler(thing, Protocol.WLAN);
        } else if (THING_TYPE_PWMLAN.equals(thingTypeUID)) {
            return new EnergeniePWMHandler(thing);
        }
```

Reply:
Done